controller/api/destination/server.go

```diff
@@ -38,153 +43,180 @@ type server struct {
+	client, _ := peer.FromContext(stream.Context())
+	log := s.log.WithFields(logging.Fields{
+		"remote": client.Addr,
+	})
+	log.Debugf("GetProfile(%+v)", dest)
+
+	host, _, err := s.getHostAndPort(dest)
+	if err != nil {
+		return err
+	}
+	service, _, err := s.getServiceAndPort(dest)
 	if err != nil {
 		return err
 	}
 
-	listener := newProfileListener(stream)
+	translator := newProfileTranslator(stream, log)
 
-	proxyNS := ""
-	parts := strings.Split(dest.GetContextToken(), ":")
-	// ns:<namespace>
-	if len(parts) == 2 && parts[0] == "ns" {
-		proxyNS = parts[1]
-	} else {
-		// TODO remove this after the 2.3 release...
-		parts = strings.Split(dest.GetContextToken(), ".")
-		// <deployment>.deployment.<namespace>.linkerd-managed.linkerd.svc.cluster.local
-		if len(parts) >= 3 {
-			log.Debug("Serving profile request for legacy proxy")
-			proxyNS = parts[2]
+	primary, secondary := newFallbackProfileListener(translator)
+
+	if ns := nsFromToken(dest.GetContextToken()); ns != "" {
+		clientProfileID := watcher.ProfileID{
+			Namespace: ns,
+			Name:      host,
 		}
-	}
-	if proxyNS != "" {
-		log.Debugf("Looking up profile given context: ns:%s", proxyNS)
-	}
 
-	err = s.resolver.streamProfiles(host, proxyNS, listener)
-	if err != nil {
-		s.log.Errorf("Error streaming profile for %s: %v", dest.Path, err)
+		s.profiles.Subscribe(clientProfileID, primary)
+		defer s.profiles.Unsubscribe(clientProfileID, primary)
 	}
-	return err
-}
 
-func (s *server) Endpoints(ctx context.Context, params *discoveryPb.EndpointsParams) (*discoveryPb.EndpointsResponse, error) {
-	s.log.Debugf("Endpoints(%+v)", params)
+	serverProfileID := watcher.ProfileID{
+		Namespace: service.Namespace,
+		Name:      host,
+	}
 
-	servicePorts := s.resolver.getState()
+	s.profiles.Subscribe(serverProfileID, secondary)
+	defer s.profiles.Unsubscribe(serverProfileID, secondary)
 
-	rsp := discoveryPb.EndpointsResponse{
-		ServicePorts: make(map[string]*discoveryPb.ServicePort),
+	select {
+	case <-s.shutdown:
+	case <-stream.Context().Done():
+		log.Debugf("GetProfile(%+v) cancelled", dest)
 	}
 
-	for serviceID, portMap := range servicePorts {
-		discoverySP := discoveryPb.ServicePort{
-			PortEndpoints: make(map[uint32]*discoveryPb.PodAddresses),
-		}
-		for port, sp := range portMap {
-			podAddrs := discoveryPb.PodAddresses{
-				PodAddresses: []*discoveryPb.PodAddress{},
-			}
-
-			for _, ua := range sp.addresses {
-				ownerKind, ownerName := s.k8sAPI.GetOwnerKindAndName(ua.pod)
-				pod := util.K8sPodToPublicPod(*ua.pod, ownerKind, ownerName)
-
-				podAddrs.PodAddresses = append(
-					podAddrs.PodAddresses,
-					&discoveryPb.PodAddress{
-						Addr: addr.NetToPublic(ua.address),
-						Pod:  &pod,
-					},
-				)
-			}
-
-			discoverySP.PortEndpoints[port] = &podAddrs
-		}
+	return nil
+}
 
-		s.log.Debugf("ServicePorts[%s]: %+v", serviceID, discoverySP)
-		rsp.ServicePorts[serviceID.String()] = &discoverySP
+func nsFromToken(ctx string) string {
```

Comment:
This logic is profiles-specific. Can this be moved behind the ProfilesWatcher interface?